### ISSUE TITLE
Meson doesn't provide a helpful error when generating coverage reports with Clang

### ISSUE THREAD

**Issue by mayl** [mesonbuild/meson#6160]
I was hitting some difficulties trying to get coverage builds to work, that I eventually realized were because meson doesn't support generating coverage builds with Clang.  What eventually tipped me off was finding [this](https://github.com/mesonbuild/meson/blob/2fe75ab31d7eb45e3615cd5bad1be6131d314dda/run_unittests.py#L5316-L5318) section in the `run_unittests.py` file.  Meson's behavior in this case was pretty confusing - `coverage-txt` and `coverage-xml` would seem to "work" but generate empty reports.  `coverage-html` would fail when `gcov` would attempt to make an enormous allocation.

I opened  #6159 to try and improve the documentation here a bit, but I think the fact that meson doesn't provide a decent error message in this case is probably still an issue.  I'll look to see if I can find a good place to add this error, but I'm not super familiar with Meson (or really Python for that matter), so it could be slow and I think this should probably be documented as an issue even if I can't fix it.

**Comment by mayl** [comment]
Ok, so I've spent about an hour trying to find an easy way to emit an error if `-Db_coverage=true` and `CC=clang` or `CXX=clang`.  So far, I haven't been able to find where to put this check, but I _think_ I have a somewhat better idea of how things happen.

At first I was looking at trying to fit the error message in `scripts/coverage.py`, but now I understand that only gets called by `ninja` after the directory is configured.  That's a little late in the process, but could still be OK.  However, I don't think it will work because I don't think `coverage.py` gets to know what compiler is in use by `ninja`.

It seems like what I need to find is the point in configuration where meson is about to write out `build.ninja`.  At that point, meson should know that `-Db_coverage=true` and which compilers are being used, and that would be a point to emit an error.  I'm just not familiar enough with meson to know where this is, can someone on the project point me towards the right file/stage to do this?

**Comment by krader1961** [comment]
It seems to me the analysis about this issue is, to this point, wrong. I've had no difficulty producing coverage reports using clang on macOS until a few days ago. That failure is due to issue #6188 which I bumped into as a consequence of upgrading my OS platform and, at the same time, updating my `meson` version to include PR #5681 caused code coverage failures.

**Comment by mayl** [comment]
Hi @krader1961.  Appreciate the analysis in your other comments (although it is slightly un-generous to me).

However, I can confirm, that after pulling the latest master (commit ffdacc4) that this is _still an issue_ on Linux with Clang.  To reproduce, I pull master and run the following commands

```
mkdir covTest && cd covTest
meson init -n testProject -l c
CC=clang meson build -Db_coverage=true
ninja -C build test
ninja -C build coverage
```

I replace meson above with `python3 <path to meson.py in git>` which I ran to ensure I was running the latest version from master.

Per this issue, the expected result of this procedure is either a working coverage report or an error.  The actual result is an empty `coverage.txt` or `coverage.xml` and a `gcov` failure when it tries to make a massive allocation.  Taking the same steps with `CC=gcc` results in success.

You aren't going to get an argument from me that coverage is working with clang in macOS (I wouldn't know) but I'm quite sure it is still broken on Linux, or the [documented procedure](https://mesonbuild.com/howtox.html) for doing this is incomplete (I submitted the note about `Clang` being unsupported).  Is there a different procedure I need to run Clang coverage reports on Linux?

I still think having the coverage build fail this way instead of communicating a clear error to the user, is bad UX, and an issue worth tracking and fixing.

**Comment by krader1961** [comment]
@mayl, Sorry, you're right in thinking my tone was a bit snippy. I had just been arguing with some ill-informed people about recent, experimental, changes to the korn shell and was not in a good mood. My apologies.

I incorporated your proposed fix into my Meson source. Your change causes a configure failure if I explicitly do `env CC=clang meson -Db_coverage=true` on macOS. I don't normally set that env var because `cc` is just a link to `clang` on macOS. On Linux (Fedora 28 in my case) coverage data is present but it's true the coverage reports contain no data using `gcovr` 4.1.2. So the problem here isn't with Meson. It's with clang on linux. The problem with your fix is that it doesn't allow for doing code coverage using clang on those platforms where it works.

I appreciate your frustration. But clang does provide coverage data that is usable on macOS with `lcov` if nowhere else. And meson does support code coverage using clang. Your proposed change is like using a cruise missile to kill a mosquito. It would be better to solve this using a more narrowly focused change. Although it's not really clear to me that any change to Meson is even warranted. Since empty coverage reports are being generated it is pretty clear there is a problem. It would be better to fix the underlying incompatibility between clang and gcovr/lcov on Linux.

**Comment by mayl** [comment]
Hi @krader1961, no worries we all get a little short sometimes.

I agree, my change was super janky, and I tried to advertise it as such, and never meant for it to be merged in as-is.  That is why I opened it as draft, and have been asking for help on how to do it better.  It was meant to demonstrate what I think meson should do in this circumstance just as a starting point.  Pointing out that it breaks working clang coverage on macOS is obviously good feedback, thanks.

On macOS is your `gcov` actually `llvm-cov gcov`? I'm obviously not an expert, but been googling around and it's not clear to me that `gnu gcov` and `llvm-cov gcov` are _meant_ to be compatible. It seems like [other](https://github.com/mesonbuild/meson/blob/2fe75ab31d7eb45e3615cd5bad1be6131d314dda/run_unittests.py#L5316-L5318) (presumably more knowledgeable) people think llvm-cov is needed as well. 

I don't know if I agree yet that this is "not a meson problem".  I could be totally off base here, but I think that if meson is going to advertise and produce a coverage build target, using hidden automatic logic, then there is a reasonable expectation that it is either going to use tools that work together, or report a configuration error.  As a novice user, when my coverage builds resulted in a gcov crash and empty reports it was not at all immediately obvious to me that I wasn't doing something wrong, or that the corrective action was to change compilers.

**Comment by krader1961** [comment]
> On macOS is your gcov actually llvm-cov gcov?

Yes. There is a `llvm-cov` command but you wouldn't normally use it, you would use */usr/bin/gcov* which is a shim to the proper program. Both report "Apple LLVM version 11.0.0".

But I don't use that as I prefer the `lcov` HTML report. So I install, what I assume is borg standard `lcov`, using Homebrew. Have you tried lcov with clang on Linux? Note that part of the problem may be due to a difference in how function definition lines are represented in the raw data. I had to create the following `lcov` wrapper script to munge the data so that `lcov` would handle it. I did not test this using borg standard gcovr. It wouldn't surprise me if installing this script as `gcovr` (and making the other obvious changes) "fixes" this issue.

```
#!/bin/sh
#
# This wraps the real lcov command so that I can deal with various bogosities
# such as LLVM (i.e., clang) not instrumenting the function definition line
# when the function is called.
#

# echo WTF: "$@" >/dev/tty
if [ "$1" = "--remove" ]
then
    /usr/local/bin/lcov "$@" '*/tests/*' '*/*.h'
    # This is needed to eliminate the function definition line from the line
    # counts since it isn't instrumented by LLVM (but is by GCC).
    # See https://github.com/linux-test-project/lcov/issues/30
    awk -F '[:,]' '
        /^SF:/ { delete defs }
        /^FN:/ { defs[$2]=1 }
        /^DA:/ { if ($3 == 0 && $2 in defs) next }
        { print }
        ' < meson-logs/coverage.info > meson-logs/coverage.info.tmp
    mv meson-logs/coverage.info.tmp meson-logs/coverage.info
else
    /usr/local/bin/lcov "$@"
fi
```

**Comment by mayl** [comment]
Ok so some progress here I think.

I actually already do use `lcov` and `genhtml`, sorry if I gave you a different impression. Unfortunately, those are not the issue - both `lcov` and `gcovr` call `gcov` under the hood and that is the issue.  What we need is for `meson` to make the coverage report use gnu `gcov` for `gcc`, and `llvm-cov gcov` for `clang`.

To test this out I replaced my `gcov` with:
```
/usr/bin/env sh
llvm-cov-6.0 gcov "$@"
```
and got a working coverage report with `clang`.  Unfortunately, leaving that mod it breaks coverage with `gcc`.  Do coverage builds with `gcc` work on your mac? 

I think I see a better path forward to getting coverage to work in both cases now.  Both `lcov` and `gcovr` have command line options that allow you to specify which `gcov` binary to use.  I think the way to do this is to have `meson` detect what compiler is in use at configuration time, find the correct `gcov` program and pass it into the `coverage.py` script.  Then I just need to modify `coverage.py` to pass the appropriate arguments to `lcov` and `gcovr` to specify the correct `gcov` binary.

I know I asked this question before, but I never really got an answer.  What is the "textbook" way to detect if clang is in use inside `meson`?

**Comment by mayl** [comment]
So some more progress on this tonight. I did a quick and dirty version most of what I proposed above in [my fork](https://github.com/mayl/meson/tree/Clang_Coverage).  Generally, except for two things it's working with both GCC and Clang:

1) I'm not detecting which compiler yet, still hoping to get some feedback on the best way to detect if clang is being used.

2) Annoyingly, because the command needed to pass into `lcov` and `gcovr` is a command with a subcommand (`llvm-cov-<version> gcov`), it is difficult to just cleanly pass into 'lcov' and 'gcovr'.  I'm working around it with my wrapper script, but this doesn't scale to other versions of `llvm-cov`.  I think we want to do this without an environment variable, but I'm struggling to think of another way to get it done.

**Comment by krader1961** [comment]
> both lcov and gcovr call gcov under the hood

Good to know. I was unaware that was the case.

> Do coverage builds with gcc work on your mac?

No idea. The only `gcc` on my system is */usr/bin/gcc* which is nothing more than a shim to `clang`. I suppose I could install the real `gcc` but I'm not that desperate :smile:

I appreciate your willingness to properly fix this rather than take the easy way out of just blacklisting `clang` since that would make me a sad puppy.

**Comment by eli-schwartz** [comment]
This was fixed by #7196
